### PR TITLE
🌱 Align ClusterClass ControlPlane with the rest of the fields

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:

This commits introduces ControlPlaneClass and allows users to set
metadata fields for all control plane objects. It also adds fields to
set infrastructure reference for Machine objects and clarifies what
contracts the field expects from the control plane provider.

NOTE: While this is technically a breaking change, it's a follow-up to #4928.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/assign @fabriziopandini @CecileRobertMichon

### CODE REVIEW

File: api/v1alpha4/clusterclass_types.go [hunk 1]
```diff
@@ -53,6 +53,20 @@ type ClusterClassSpec struct {
 	Workers WorkersClass `json:"workers,omitempty"`
 }
 
+type ControlPlaneClass struct {
```
Comment (sbueringer) [inline api/v1alpha4/clusterclass_types.go:56]: Should we update the proposal accordingly?

File: api/v1alpha4/clusterclass_types.go [hunk 2]
```diff
@@ -53,6 +53,20 @@ type ClusterClassSpec struct {
...
+
+	// LocalObjectTemplate contains the reference to the control plane provider.
+	LocalObjectTemplate `json:",inline"`
+
+	// MachineTemplate defines the metadata and infrastructure information
+	// for control plane machines.
+	//
+	// This field is supported if and only if the control plane provider template
+	// referenced above is Machine based and supports setting replicas.
+	MachineInfrastructure LocalObjectTemplate `json:"machineInfrastructure"`
```
Comment (fabriziopandini) [inline api/v1alpha4/clusterclass_types.go:67]: @sbueringer @killianmuldoon we should update the definition of cluster class compatibility considering this field
@ykakarap we should consider this in additional ClusterClass validations

**Comment by vincepri** [comment]
@fabriziopandini This should be good to go

File: api/v1alpha4/clusterclass_types.go (hunk 1 above)
Comment (vincepri) [reply api/v1alpha4/clusterclass_types.go:57]: Yes, I can do that in here

File: api/v1alpha4/clusterclass_types.go [hunk 3]
```diff
@@ -53,6 +53,23 @@ type ClusterClassSpec struct {
 	Workers WorkersClass `json:"workers,omitempty"`
 }
 
+// ControlPlaneClass defines the class for the control plane.
+type ControlPlaneClass struct {
+	Metadata ObjectMeta `json:"metadata,omitempty"`
+
+	// LocalObjectTemplate contains the reference to the control plane provider.
+	LocalObjectTemplate `json:",inline"`
```
Comment (ykakarap) [inline api/v1alpha4/clusterclass_types.go:61]: I am probably missing something but the `LocalObjectTemplate` we have here refs to a ControlPlaneTemplate (Ex: KubeadmControlPlaneTemplate - will be introduced in https://github.com/kubernetes-sigs/cluster-api/pull/4904 ).

This ControlPlaneTemplate will internally refer to a machine infrastructure, isnt it? 
Going back the kcp template example we have a machine infra ref at `kcptemplate.spec.template.spec.machineTemplate.infrastructureRef`.   How is this different from the new `MachineInfrastructure` field?

Comment (sbueringer) [reply api/v1alpha4/clusterclass_types.go:61]: Imho both will reference the same InfrastructureMachineTemplate. This PR makes it explicit on the ClusterClass (i.e. you have to set the ref on the ClusterClass **and** inside the ControlPlaneTemplate)

Comment (fabriziopandini) [reply api/v1alpha4/clusterclass_types.go:61]: Exactly, we are making this explicit;
Also, according to the initial work on https://github.com/kubernetes-sigs/cluster-api/pull/5002, ClusterClass is always going to set the reference inside the ControlPlaneTemplate, so it is not required to set it in both places upfront

Comment (sbueringer) [reply api/v1alpha4/clusterclass_types.go:61]: Good that fits for my PR as I assume that the ref can be unset in the ControlPlane even if it is set in the ClusterClass.

Comment (ykakarap) [reply api/v1alpha4/clusterclass_types.go:61]: >ClusterClass is always going to set the reference inside the ControlPlaneTemplate, so it is not required to set it in both places upfront

Good to know that we are doing this, however the `kcptemplate.spec.template.spec.machineTemplate.infrastructureRef` is not an optional field. So are we expecting that the user set the filed **both** inside the ControlPlaneTemplate and inside the ClusterClass definition and also that both the fields point to the same thing?
We should throw an error incase they dont match, is it?

**Pull request merged**